presto-main/src/main/java/com/facebook/presto/memory/LowMemoryTaskKiller.java

```diff
@@ -0,0 +1,187 @@
+    private final NotificationListener gcNotificationListener = (notification, ignored) -> onGCNotification(notification);
+    private final SqlTaskManager sqlTaskManager;
+    private final boolean lowMemoryTaskKillerMemoryEnabled;
+    private final double lowMemoryTaskKillerMemoryThreshold;
+    private final DataSize lowMemoryTaskKillThreshold;
+
+    @Inject
+    public LowMemoryTaskKiller(SqlTaskManager sqlTaskManager, TaskManagerConfig taskManagerConfig)
+    {
+        this.sqlTaskManager = requireNonNull(sqlTaskManager, "sqlTaskManager must not be null");
+        this.lowMemoryTaskKillerMemoryEnabled = taskManagerConfig.isLowMemoryTaskKillerEnabled();
+        this.lowMemoryTaskKillerMemoryThreshold = taskManagerConfig.getLowMemoryTaskKillerMemoryThreshold();
+        this.lowMemoryTaskKillThreshold = taskManagerConfig.getLowMemoryTaskKillerThreshold();
+    }
+
+    @PostConstruct
+    public void start()
+    {
+        if (!lowMemoryTaskKillerMemoryEnabled) {
+            return;
+        }
+
+        for (GarbageCollectorMXBean mbean : ManagementFactory.getGarbageCollectorMXBeans()) {
+            if (mbean.getName().equals("TestingMBeanServer")) {
+                continue;
+            }
+
+            ObjectName objectName = mbean.getObjectName();
+            try {
+                ManagementFactory.getPlatformMBeanServer().addNotificationListener(
+                        objectName,
+                        gcNotificationListener,
+                        null,
+                        null);
+            }
+            catch (JMException e) {
+                throw new RuntimeException("Unable to add listener", e);
+            }
+        }
+    }
+
+    @PreDestroy
+    public void stop()
+    {
+        if (!lowMemoryTaskKillerMemoryEnabled) {
+            return;
+        }
+
+        for (GarbageCollectorMXBean mbean : ManagementFactory.getGarbageCollectorMXBeans()) {
+            ObjectName objectName = mbean.getObjectName();
+            try {
+                ManagementFactory.getPlatformMBeanServer().removeNotificationListener(objectName, gcNotificationListener);
+            }
+            catch (JMException ignored) {
+                log.error("Error removing notification: " + ignored);
+            }
+        }
+    }
+
+    private void onGCNotification(Notification notification)
+    {
+        if (GC_NOTIFICATION_TYPE.equals(notification.getType())) {
+            GarbageCollectionNotificationInfo info = new GarbageCollectionNotificationInfo((CompositeData) notification.getUserData());
+            if (info.isMajorGc()) {
+                DataSize beforeGcDataSize = info.getBeforeGcTotal();
+                DataSize afterGcDataSize = info.getAfterGcTotal();
+                long garbageCollectedBytes = beforeGcDataSize.toBytes() - afterGcDataSize.toBytes();
+
+                if (isLowMemory() && garbageCollectedBytes < lowMemoryTaskKillThreshold.toBytes()) {
```

Comment:
Should we have some kind of a "quiet period" after it has killed a query to observe the effects ? Like how do we prevent this from going rogue ?

Reply:
Full GC won't happen once the query has been killed unless there are more memory hungry queries running and consuming memory resulting in the same scenario. And in that case we want it to kick in and kill them to avoid jvm oom.